### ISSUE TITLE
duckdb raise the error too late when create index with an exists name

### ISSUE THREAD

**Issue by l1t1** [duckdb/duckdb#9740]
### What happens?

when create an index with an exists name, it took many time to create it first, then report the error.
it should check the name first.

### To Reproduce

```sql
create table sc 
  ( 
     s_id    int, 
     c_id    int, 
     t_id    int,
     score int 
  ) ;

insert into sc select i,floor(random()*1000),floor(random()*10000000),floor(random()*750)from range(1,50000001)t(i) ;
create index idx_s_id  on sc(s_id)   ;

create index idx_s_id  on sc(s_id)   ;
```
```
 33% ▕███████████████████▊                                        ▏ Run Time (s): real 14.013 user 36.925437 sys 40.544660
Error: Catalog Error: Index with name "idx_s_id" already exists!
```

### OS:

windows7 x64

### DuckDB Version:

0.9.2

### DuckDB Client:

CLI

### Full Name:

lutao

### Affiliation:

study

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by kevinmingtarja** [comment]
Hi, I'm interested in contributing to this issue.

**Comment by kevinmingtarja** [comment]
Hi, I managed to do some profiling on the execution and have a better view on the cause, but I may need some guidance to move forward.

I profiled using [xctrace](https://keith.github.io/xcode-man-pages/xctrace.1.html) as I'm on macos. And I used the same steps to reproduce, which is creating a table with millions of rows, then do `create index` statements on an existing index.

I profiled a single execution of a duplicate `create index` statement. This particular execution took `69.04s` of CPU time in total, but this is spread across 8 threads (not fully shown in the screenshot). I found that 7 of the threads each spent around `5.5 to 6.0s` on `PhysicalCreateARTIndex::Sink`, so if we multiply this by 7, this takes up more than half of the total CPU time. 

<img width="985" alt="Screen Shot 2023-12-05 at 13 13 16" src="https://github.com/duckdb/duckdb/assets/69668484/69e3cf59-c8ce-4e0c-a137-b806ce403fb7"><br/><br/>

Then the 1 other thread that's not doing this is doing `PhysicalCreateARTIndex::Finalize` and spent `6.30s` just to traverse and verify the index. The call to `DuckSchemaEntry::CreateIndex`, which is where the checking of duplicate index is done and throws the CatalogException, only happens after this index verification step.

<img width="985" alt="Screen Shot 2023-12-05 at 13 22 49" src="https://github.com/duckdb/duckdb/assets/69668484/7dccc1fa-c399-4ca7-ba97-9130dccae04c"><br/><br/>

So, looking at this, should we move the checking of the name out of `PhysicalCreateARTIndex`, to avoid the heavy work of materializing all the data and traversing and verifying it?

`.trace` file for reference: [trace.zip](https://github.com/duckdb/duckdb/files/13555439/trace.zip)
